wire-ios-data-model/WireDataModel.docc/encryption-at-rest.md

```diff
@@ -0,0 +1,76 @@
+
+The stored update events are decrypted with the associated private key. If this is the primary key, then the event can only be decrypted (and subsequently processed) when the app is in the foreground with user presence. If it is the secondary key, the event can be decrypted and processed without user presence in the background.
+
+For this reason, only call events are encrypted and decrypted using the secondary keys, since these events need to be processed while the app is operating in the background during CallKit calls. 
+
+Key | Location | Accessible | Purpose                          
+--- | -------- | ------ | ---------------------- 
+Primary public | Secure Enclave | Always | Encrypt all stored update events (except call events) and the database key.
+Primary secondary | Secure Enclave | User presence | Decrypt all stored update events (except call events) and the database key.
+Secondary public | Secure Enclave | Always | Encrypt call events only.
+Secondary private | Secure Enclave | After first unlock | Decrypt call events only.
+Database | Keychain | User presence | Encrypt and decrypt all messages.
+
+### User flow
+
+To illustrate how all these pieces work together, consider a user who wishes to view a message they have received while encryption at rest is enabled:
+
+1. First, the user opens the app.
+2. The user is prompted to verify their identity via biometric authentication.
+3. If successful, the app uses the authentication result to fetch the private keys from the Secure Enclave (via the Keychain).
+4. The app also fetches the encrypted database key from Keychain.
+5. The database key is decrypted and stored in memory. At this point, the database is considered unlocked.
+6. A new message is received, then encrypted with the database key, then stored in the database.
+7. The user wishes to view the message in the UI.
+8. The encrypted message is retrieved from the database, decrypted with the database key, then presented to the user.
+9. The user closes the app.
+10. The database key is discared, the database is considered locked.
+
+![Basic flow describing user unlocking database to access messages](encryption-at-rest-in-action)
+
+### Data migrations
+
+#### When enabling / disabling encryption at rest
+
+When encryption at rest is turned on a migration is needed to bring the existing database into the correct state. Once the encryption keys have been generated, we then proceed to encrypt all messages in the database with the database key. Similarly, when encryption at rest is turned off, we must decrypt all messages with the database key before the encryption keys are destroyed, otherwise the messages no longer be accessible.
```

Comment:
```suggestion
When encryption at rest is turned on a migration is needed to bring the existing database into the correct state. Once the encryption keys have been generated, we then proceed to encrypt all messages in the database with the database key. Similarly, when encryption at rest is turned off, we must decrypt all messages with the database key before the encryption keys are destroyed, otherwise the messages will no longer be accessible.
```